### PR TITLE
Support error formats that require all annotations to print

### PR DESCRIPTION
**Pull request by srikrsna-buf**
Related to #829 

The current approach formats file annotations one after the other joining them with a new line. This works for the current set of formats (text, msvs, and json). However, it falls short when trying to print in a format such as `jUnit`. Here is a sample `jUnit` xml:

```xml
<testsuite name="buf-lint" tests="3" failures="3" errors="0">
    <testcase name="foo/foo.proto">
        <failure type="PACKAGE_VERSION_SUFFIX">Package name "foo" should be suffixed with a correctly formed version, such as "foo.v1"</failure>
        <failure type="MESSAGE_PASCAL_CASE">Message name "snake_case_message" should be PascalCase, such as "SnakeCaseMessage"</failure>
        <failure type="FIELD_LOWER_SNAKE_CASE">Field name "CamelCaseField" should be lower_snake_case, such as "camel_case_field"</failure>
    </testcase>
</testsuite>
```
To support such formats, this PR refactors `bufanalysis` package to use a new `printer` interface. Each `Format` gets a printer. This removes methods used for formatting from `FileAnnotation` interface. All other public api is unchanged.

### CODE REVIEW

File: private/bufpkg/bufanalysis/bufanalysis.go
Comment (doriable) [inline private/bufpkg/bufanalysis/bufanalysis.go:71]: Instead of having this be a global, we could rework `FormatFileAnnotation` with the `switch` statement to instead operate on the entire set of file annotations, `[]FileAnnotation`?

The concern is for the `junit` implementation, where we'll need all the `FileAnnotations` to print, so the print logic would need to be one level up from the single `FileAnnotation` unit, so by reworking the logic such that we handle the entire slice of `[]FileAnnotation` given the `formatString` provided to `PrintAllFileAnnotations` should be sufficient.

File: private/bufpkg/bufanalysis/bufanalysis.go
Comment (bufdev) [inline private/bufpkg/bufanalysis/bufanalysis.go:199]: Need to check to make sure printer from map is not nil, and if it is, error

File: private/bufpkg/bufanalysis/bufanalysis.go
Comment (bufdev) [inline private/bufpkg/bufanalysis/bufanalysis.go:210]: You can't ignore this error, you need to propagate it

File: private/bufpkg/bufimage/bufimagebuild/builder_test.go
Comment (bufdev) [inline private/bufpkg/bufimage/bufimagebuild/builder_test.go:377]: check the error with `assert.NoError(t, err)`

File: private/bufpkg/bufanalysis/bufanalysis.go
Comment (srikrsna-buf) [reply private/bufpkg/bufanalysis/bufanalysis.go:210]: Even if `printFileAnnotationAsText` always returns `nil` similar to `bytes.Buffer`?

File: private/bufpkg/bufanalysis/bufanalysis.go
Comment (srikrsna-buf) [reply private/bufpkg/bufanalysis/bufanalysis.go:71]: I think we can remove `FormatFileAnnotation` altogether, it is only being used in tests. 

IMO separating the printing part from `FileAnnotation` is useful as the formats can grow. Instead of modifying the `FileAnnotation` interface we can just add a new `Format` and `printer`. 

The `FileAnnotation` can have a default format in the form of `Stringer` interface (used in tests of `bufimagebuild`).

Not that it is major concern here, but this is slightly faster than using `FormatFileAnnotation` to return a `string` as we can reuse the `bytes.Buffer`

Comment (doriable) [reply private/bufpkg/bufanalysis/bufanalysis.go:71]: So I think that we probably want to prevent `Format` from becoming more than just the enum for managing format types (which this current design adheres to), however, we should also avoid this global here if it can be handled by a `switch`. We should also avoid attaching the `printer` interface to the `Format`, even though they are 1:1 since that logic can be encapsulated into a single `Print` function that takes in the format and file annotations (without defining a new structure). I agree with the decision to probably get rid of `FormatFileAnnotation`, since for the `junit` output in particular, we will need to print all `FileAnnotation`s together, so formatting each file annotation separately doesn't really make much sense anymore.

**Review by doriable — APPROVED** [review]
Looks good to me, and I also pulled it and did a couple of tests against this :)

File: private/bufpkg/bufanalysis/bufanalysistesting/bufanalysistesting_test.go
Comment (bufdev) [inline private/bufpkg/bufanalysis/bufanalysistesting/bufanalysistesting_test.go:55]: You don't need to do this, you can just do:

```
`This
is how you do multiline strings
and more`
```

File: private/bufpkg/bufanalysis/bufanalysis.go
Comment (bufdev) [inline private/bufpkg/bufanalysis/bufanalysis.go:104]: Should we kill this as well?

File: private/bufpkg/bufanalysis/bufanalysistesting/bufanalysistesting_test.go
Comment (srikrsna-buf) [reply private/bufpkg/bufanalysis/bufanalysistesting/bufanalysistesting_test.go:55]: Tried that, gofmt is formatting it like this,
```go
	assert.Equal(
		t,
		`path/to/file.proto:1:1:Hello.
path/to/file.proto:2:1:Hello.
`,
		sb.String(),
	)
```

File: private/bufpkg/bufanalysis/bufanalysis.go
Comment (srikrsna-buf) [reply private/bufpkg/bufanalysis/bufanalysis.go:104]: Stringer is useful (in one test so far) and generic enough (not an encoding format) to be part of the interface. I think we should let this be.

File: private/bufpkg/bufanalysis/bufanalysistesting/bufanalysistesting_test.go
Comment (bufdev) [reply private/bufpkg/bufanalysis/bufanalysistesting/bufanalysistesting_test.go:55]: Yep thats right

File: private/bufpkg/bufanalysis/bufanalysis.go
Comment (bufdev) [reply private/bufpkg/bufanalysis/bufanalysis.go:104]: It's saying that this is how you do the text format though - we should at least update the comment to not reference the text format and not tie it to the text format

Comment (srikrsna-buf) [reply private/bufpkg/bufanalysis/bufanalysis.go:104]: Updated the comment

File: private/bufpkg/bufanalysis/bufanalysis.go
Comment (bufdev) [reply private/bufpkg/bufanalysis/bufanalysis.go]: If it will always return nil, then the function should not have an error signature. I think what you mean is that writes to the `buffer` will never result in error, so this function, due to how we happen to be calling it, will always return nil. That's a big usage assumption IMO - if we later change from `bytes.Buffer` to something else that can error, we are then ignoring an error. We should always check errors such as these.

**Comment by bufdev** [comment]
Do we still want to pursue this?

**Comment by doriable** [comment]
> Do we still want to pursue this?

From what I can see, after merging this PR, the bulk of the remaining work would be implementing a junit printer, with the above proposed layout (which @srikrsna-buf has solicited the opinions of other team members and the general consensus seems like this is a sane test output for junit).

I think this should be pretty quick to close out, but I know there's also a lot on @srikrsna-buf's plate right now. If he'd like to close it out, I'll let him comment, otherwise, I'm happy to add it to my to-do list.

**Comment by srikrsna-buf** [comment]
@doriable Thank you for the response and the offer. We can merge this without any changes. 

I'll implement the `juit` after this. Although it will not be my priority for at least 2 weeks. Feel free to take it up @doriable

**Comment by srikrsna-buf** [comment]
@doriable Thank you for the response and the offer. We can merge this without any changes. 

I'll implement the `juit` after this. Although it will not be my priority for at least 2 weeks. Feel free to take it up @doriable

**Pull request merged**
